Gen cesm catalog

mnlevy1981 opened the pull request:
Added a script that generates an intake catalog for time series data generated by CESM if `pyReshaper` was run in post-processing.

```
$ ./gen_CESM_catalog.py -c /glade/p/cgd/oce/people/mlevy/cases/b.e22b05.B1850.f09_g17.timeseries_output_for_intake/
INFO (_gen_timeseries_catalog): Will catalog files in /glade/p/cgd/oce/people/mlevy/archive/b.e22b05.B1850.f09_g17.timeseries_output_for_intake
INFO (_gen_timeseries_catalog): Creating /glade/p/cgd/oce/people/mlevy/archive/b.e22b05.B1850.f09_g17.timeseries_output_for_intake/intake/cesm_catalog.csv.gz...
```

```
$ zcat /glade/p/cgd/oce/people/mlevy/archive/b.e22b05.B1850.f09_g17.timeseries_output_for_intake/intake/cesm_catalog.csv.gz | head -n 4
case,component,stream,variable,start_date,end_date,path,parent_branch_year,child_branch_year,parent_case
b.e22b05.B1850.f09_g17.timeseries_output_for_intake,atm,cam.h0,TAUBLJY,000101,000112,../atm/proc/tseries/month_1/b.e22b05.B1850.f09_g17.timeseries_output_for_intake.cam.h0.TAUBLJY.000101-000112.nc,-1,-1,-
b.e22b05.B1850.f09_g17.timeseries_output_for_intake,atm,cam.h0,num_c2SFWET,000101,000112,../atm/proc/tseries/month_1/b.e22b05.B1850.f09_g17.timeseries_output_for_intake.cam.h0.num_c2SFWET.000101-000112.nc,-1,-1,-
b.e22b05.B1850.f09_g17.timeseries_output_for_intake,atm,cam.h0,dst_c3,000101,000112,../atm/proc/tseries/month_1/b.e22b05.B1850.f09_g17.timeseries_output_for_intake.cam.h0.dst_c3.000101-000112.nc,-1,-1,-
```

Note that this file assumes `intake-esm` can handle a relative path from the `csv.gz` file to the netCDF data.

I can think of several improvements this script needs, some which might belong in this PR and others that might spawn new issue tickets.

1. Error handling if file name doesn't fit the `{casename}.{stream}.{variable}.{start_date}-{end_date}.nc` template (e.g. if history files and time series are collocated, which is default behavior of CESM postprocessing)
1. Backup plan for determining location of time series if `pp_config` is not available
1. I don't think

   ```
    catalog['parent_branch_year'] = entry_cnt*[run_config['RUN_REFDATE']]
    catalog['child_branch_year'] = entry_cnt*[run_config['RUN_STARTDATE']]
   ```

   will work for determining branch point if a run is based off a reference case, as they aren't in `run_config` yet

And I'm sure additional issues will come up, but I wanted to open this PR to advertise that this script is working in tightly controlled instances.

Closes #2

mnlevy1981 committed:
Script that queries CASEROOT for some info

First pass just prints to screen data needed from xml files

mnlevy1981 committed:
Update how data is stored in memory

mnlevy1981 committed:
Generate catalog from time series!

andersy005 commented on cesmcatalog/gen_CESM_catalog.py:
```diff
@@ -0,0 +1,170 @@
+    sys.exit(1)
+
+  run_config = dict()
+  for var in ['GET_REFCASE', 'RUN_REFCASE']:
+    run_config[var] = subprocess.check_output('./xmlquery --value {}'.format(var), shell=True)
+  DOUT_S = subprocess.check_output('./xmlquery --value DOUT_S', shell=True)
+  if DOUT_S == 'TRUE':
+    DOUT_S_ROOT = subprocess.check_output('./xmlquery --value DOUT_S_ROOT', shell=True)
```
To be safe, we may need some error handling here by wrapping these lines a try block. What do you think?

mnlevy1981 replied:
Yeah, error handling would be great. I'll add it to the list :)

mnlevy1981 replied:
Now that we're relying on the `CIME.Case` object instead of running these via `subprocess`, the failure mode is returning `None` rather than the expected variable value; that's worth checking for, but it'll be an `if` statement instead of a `try` / `except` block.

andersy005 commented on cesmcatalog/gen_CESM_catalog.py:
```diff
@@ -0,0 +1,170 @@
+################################################################################
+
+def _gen_timeslice_catalog(case_root, archive_root, run_config):
+  # TODO: figure out how to generate catalog from time slice
```
Unless there is extra information saved in the xml files, we need to open the files to figured out what variables are present, and harvest additional metadata if necessary. https://github.com/NCAR/intake-esm-datastore/pull/70 is a good fit for this

mnlevy1981 replied:
In 821895e I used `xarray` to open the netcdf files and pull out `long_name`; now instead of taking a second and a half to generate a catalog (from a single year of a fully-coupled run), it's taking a minute and a half... I do cache the long name for each variable, so we only open one file per variable (even if the full time series is split across many files).

mnlevy1981 committed:
Should pass black checks now

mnlevy1981 committed:
Should pass flake8 checks now

mnlevy1981 committed:
Should pass isort checks now

mnlevy1981 committed:
Updated wording on generation log message

I didn't like having {filename}... because it looked like we were creating a
.csv.gz... file

mnlevy1981 committed:
Added RUN_REFDATE and RUN_STARTDATE to run_config

Also, first time running with the pre-commit hooks picked up some formatting
changes (unclear why this didn't happen in the CI framework)

mnlevy1981 committed:
Get CASE from xml file

Only look for {CASE}*.nc instead of *.nc

mnlevy1981 committed:
Add long_name to the catalog

This entails opening netCDF files to get the long_name attribute; the current
implementation opens one file per variable name per component, but if a
variable is spread across multiple files it is assumed that the long_name does
not change.

Also created local copies of some of the other data in the catalog to make it
easier to reference between columns (e.g. storing path locally so I don't need
to access catalog['path'][-1] to get the most recent path value)

mnlevy1981 committed:
Add --debug option

Right now, the script itself does not use the debug message level but running
with -d will add some environment information to the output.

andersy005 commented on cesmcatalog/gen_CESM_catalog.py:
```diff
@@ -0,0 +1,224 @@
+    logger = logging.getLogger(__name__)
+
+    # 1. change directory to case_root (if it exists)
+    try:
+        os.chdir(case_root)
+    except:
+        # TODO: set up logger instead of print statements
+        logger.error('{} does not exist'.format(case_root))
+        sys.exit(1)
```
We may want to replace this with a contextmanager approach:

```python
from contextlib import contextmanager
@contextmanager
def chdir(path):
    """
    Change working directory to `path` and restore it again
    This context manager is useful if `path` stops existing during your
    operations.
    """
    old_dir = os.getcwd()
    os.chdir(path)
    try:
        yield
    finally:
        os.chdir(old_dir)
```

You can then use it as follows:

```python
with chdir(case_root):
   ... DO SOME WORK
```

mnlevy1981 replied:
Nice, I'll look into this tomorrow (comment is "outdated" but hasn't been resolved yet)

andersy005 commented on cesmcatalog/gen_CESM_catalog.py:
```diff
@@ -0,0 +1,224 @@
+            if component not in long_name_dict:
+                long_name_dict[component] = dict()
+            if variable not in long_name_dict[component]:
+                ds = xr.open_dataset(path, decode_cf=False)
```
Using dask by specifying `chunks={}` may improve performance:

```python
In [1]: import xarray as xr

In [2]: %time xr.tutorial.open_dataset('rasm', decode_times=False)
CPU times: user 6.08 ms, sys: 9.71 ms, total: 15.8 ms
Wall time: 452 ms

In [3]: %time xr.tutorial.open_dataset('rasm', decode_times=False, chunks={})
CPU times: user 3.29 ms, sys: 873 µs, total: 4.17 ms
Wall time: 17.8 ms
```

```suggestion
                ds = xr.open_dataset(path, decode_cf=False, chunks={})
```

mnlevy1981 replied:
Done via e091854d, thanks! I'm not seeing a performance boost yet, but I wonder if it'll speed things up once we get to bigger time series files.

jedwards4b commented on cesmcatalog/gen_CESM_catalog.py:
```diff
@@ -0,0 +1,224 @@
+################################################################################
+
+
+def _find_data(case_root):
```
Instead of reinventing stuff why not interface the cime tools that already exist?

jedwards4b commented on cesmcatalog/gen_CESM_catalog.py:
```diff
@@ -0,0 +1,224 @@
+        sys.exit(1)
+
+    # 2. Collect info on where time slice output is
+    if not os.path.isfile('./xmlquery'):
```
Instead of using xmlquery I think that you should use cime objects directly, for example:

  with Case(first_case, read_only=False) as case:
       run_refcase = case.get_value('RUN_REFCASE')

mnlevy1981 replied:
I'd be happy to use the `Case` class, but struggled to figure out how to access it. Before I can import `Case` I need to know `CIMEROOT` right? Or are the python libraries symlinked in case root?

```
sys.path.append(os.path.join(subprocess.check_output('./xmlquery --value CIMEROOT', shell=True), 'scripts', 'lib'))
from CIME.case import Case
```

works, but I assume there's a cleaner way?

jedwards4b replied:
There are a number of ways to do it, the below uses an environment variable.  In cime we symlink the files in the case directory to their location in the cime source tree which also works (for example xmlquery). 

```
cimeroot = os.environ.get('CIMEROOT')
_LIBDIR = os.path.join(cimeroot, "scripts", "Tools")                                                                         
sys.path.append(_LIBDIR)                                                                                                             
                                                                                                                                     
from standard_script_setup          import *                                                                                         
                                                                                                                                     
from CIME.case import Case                                                                                                           
                                                                                                                                     
logger = logging.getLogger(__name__)                        
```

mnlevy1981 replied:
I added an optional `--cimeroot` option to the script, and if it isn't available then I call `xmlquery` once to get it added to the path, but the rest is much cleaner:

```
    run_config = dict()
    with Case(caseroot, read_only=True) as case:
        for var in ['CASE', 'GET_REFCASE', 'RUN_REFCASE', 'RUN_REFDATE', 'RUN_STARTDATE']:
            run_config[var] = case.get_value(var)

        if case.get_value('DOUT_S') == 'TRUE':
            DOUT_S_ROOT = case.get_value('DOUT_S_ROOT')
        else:
            DOUT_S_ROOT = None
```

Once this script is called by CESM postprocessing, we'll either use `--cimeroot` or I'll modify it to check the environment variable. Thanks for the pointers here!

mnlevy1981 committed:
Get CIME variables from Case object

Still use xmlquery to get CIMEROOT, which is needed to import Case.

mnlevy1981 committed:
Drop "import *" and make case read_only

mnlevy1981 committed:
Add optional --cimeroot argument

If we already know cimeroot, no need for xmlquery at all -- this will be useful
once this script is part of the post-processing suite, but for now we fallback
to using xmlquery to set up the path to the CIME python libraries.

jedwards4b reviewed (approved):
Looks good, thanks.
